File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.js
// store turnaround time, createdAt, and index in priority queue
// While iterating over time:
//   Initialize the current job if now === createdAt + turnaroudtime
//   Enqueue a job in the priority queue if now === createdAt
//   Dequeue a new job and set now to it when current job !== null

class PriorityQueue {
    constructor() {
        this.heap = [];
    }
    
    get size() {
        return this.heap.length;
    }
    
    get isEmpty() {
        return this.size === 0;
    }
    
    getParent(index) {
        return Math.floor((index - 1) / 2);
    }
    
    getChildren(index) {
        return [ 2 * index + 1, 2 * index + 2 ];
    }
    
    swap(i, j) {
        [ this.heap[i], this.heap[j] ] = [ this.heap[j], this.heap[i] ];
    }
    
    compare(node1, node2) {
        if (node1.turnaround - node2.turnaround < 0) return true;
        if (node1.turnaround - node2.turnaround === 0 && node1.createdAt - node2.createdAt < 0) return true;
        if (node1.turnaround - node2.turnaround === 0 && node1.createdAt - node2.createdAt === 0 && node1.index - node2.index < 0) return true;
        return false;
    }
    
    bubbleUp() {
        let index = this.size - 1;
        
        while (index > 0) {
            const parent = this.getParent(index);
            if (this.compare(this.heap[parent], this.heap[index])) break;
            this.swap(index, parent);
            index = parent;
        }
    }
    
    enqueue(element) {
        this.heap.push(element);
        if (this.size === 1) return;
        this.bubbleUp();
        
    }
    
    bubbleDown() {
        let index = 0;
        
        while (index < this.size) {
            let minimum = index;
            const [ left, right ] = this.getChildren(index);
            if (left < this.size && this.compare(this.heap[left], this.heap[minimum])) minimum = left;
            if (right < this.size && this.compare(this.heap[right], this.heap[minimum])) minimum = right;
            if (minimum === index) break;
            this.swap(index, minimum);
            index = minimum;
        }
        
    }
    
    dequeue() {
        if (this.isEmpty) return null;
        if (this.size === 1) return this.heap.pop();
        if (this.size === 2) {
            this.swap(0, this.size - 1);
            return this.heap.pop();
        }
        
        this.swap(0, this.size - 1);
        const node = this.heap.pop();
        this.bubbleDown();
        return node;
    }
}

function solution(jobs) {
    const requests = jobs.map(([ createdAt, turnaround, ], index) => ({
        turnaround,
        createdAt,
        index
    }));
    
    const sortedRequests = requests.sort(({ createdAt: createdAt1 }, { createdAt: createdAt2 }) => createdAt1 - createdAt2);
    const priorityQueue = new PriorityQueue();
    
    let upcomingRequestIndex = 0;
    let currentRequest = null;
    let averageTimeLapse = 0;
    
    for (let now = 0; now < 1_000_0000; now++) {
        if (currentRequest && currentRequest.turnaround === 0) {
            averageTimeLapse += (now - currentRequest.createdAt);
            currentRequest = null;
        };
        while (upcomingRequestIndex < sortedRequests.length && sortedRequests[upcomingRequestIndex].createdAt === now) {
            priorityQueue.enqueue(sortedRequests[upcomingRequestIndex]);
            upcomingRequestIndex += 1;
        }
        if (!currentRequest && !priorityQueue.isEmpty) {
            currentRequest = priorityQueue.dequeue();
        }
        if (currentRequest) currentRequest.turnaround -= 1;
        
    }
    
    averageTimeLapse = Math.floor(averageTimeLapse / sortedRequests.length);
    return averageTimeLapse;
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq
from collections import deque

def solution(jobs):
    todos = deque(sorted([ [ jobs[index][1], jobs[index][0], index ] for index in range(0, len(jobs)) ], key=lambda todo: todo[1]))
    
    # 소요 시간 오름차순, 요청 시각 오름차순, 번호 오름차순
    
    queue = []
    current = None
    trunaround_history = []
    
    time = 0
    
    while len(queue) > 0 or len(todos) > 0 or current != None:
        while len(todos) > 0 and todos[0][1] == time:
            heapq.heappush(queue, todos.popleft())
            
            
        if current != None:
            current[0] -= 1
            if current[0] == 0:
                trunaround_history.append(time - current[1])
                current = None if len(queue) == 0 else heapq.heappop(queue)
        else:
            current = None if len(queue) == 0 else heapq.heappop(queue)

        
        time += 1
        
    
    
    return sum(trunaround_history) // len(trunaround_history)
    



# 대기 큐에 있거나 할일에 있으면
#  현재 시점에 요청된 작업 push
#  현재 실행중인 작업의 남은 시간 -= 1
#  현재 실행중인 작업의 남은 시간 === 0
#    history에 기록(작업 종료 시각, 즉 time)
#    대기 큐에서 우선순위가 높은 작업 꺼내서 현재 실행중인 작업으로 할당 (작업 소요시간 작은 것, 작업 요청 시각이 빠른 것, 작업 번호가 작은 것)
#  time += 1